provider/datagen/src/lib.rs

```diff
@@ -309,36 +310,59 @@ impl DatagenProvider {
         &self,
         key: DataKey,
         locale: &DataLocale,
-        fallbacker: &LocaleFallbacker,
+        fallbacker: &Lazy<
+            Result<LocaleFallbacker, DataError>,
+            impl FnOnce() -> Result<LocaleFallbacker, DataError>,
+        >,
     ) -> Result<Option<DataPayload<ExportMarker>>, DataError> {
         log::trace!("Generating key/locale: {key}/{locale:}");
         let mut metadata = DataRequestMetadata::default();
         metadata.silent = true;
-        let config = LocaleFallbackConfig::from_key(key);
-        let mut iter = fallbacker.for_config(config).fallback_for(locale.clone());
+        // Lazy-compute the fallback iterator so that we don't always require CLDR data
+        let mut option_iter: Option<Result<LocaleFallbackIterator, DataError>> = None;
         loop {
             let req = DataRequest {
-                locale: iter.get(),
+                locale: match option_iter.as_ref() {
+                    Some(iter) => {
+                        let iter = iter.as_ref().map_err(|x| *x)?;
+                        iter.get()
+                    }
+                    None => locale,
+                },
                 metadata,
             };
             let result = self.load_data(key, req);
             match result {
                 Ok(data_response) => {
                     #[allow(clippy::unwrap_used)] // LocaleFallbackProvider populates it
-                    if iter.get().is_empty() && !locale.is_empty() {
-                        log::debug!("Falling back to und: {key}/{locale}");
+                    if let Some(iter) = option_iter.as_ref() {
+                        let iter = iter.as_ref().map_err(|x| *x)?;
+                        if iter.get().is_empty() && !locale.is_empty() {
+                            log::debug!("Falling back to und: {key}/{locale}");
+                        }
                     }
                     return Ok(Some(data_response.take_payload()?));
                 }
                 Err(DataError {
                     kind: DataErrorKind::MissingLocale,
                     ..
                 }) => {
-                    if iter.get().is_empty() {
-                        log::debug!("Could not find data for: {key}/{locale}");
-                        return Ok(None);
-                    }
-                    iter.step();
+                    match option_iter.as_mut() {
+                        Some(iter) => {
+                            let iter = iter.as_mut().map_err(|x| *x)?;
+                            if iter.get().is_empty() {
+                                log::debug!("Could not find data for: {key}/{locale}");
+                                return Ok(None);
+                            }
+                            iter.step();
+                        }
+                        None => (),
+                    };
+                    option_iter.get_or_insert_with(|| {
```

Comment:
You can return the error here if you don't use `get_or_insert_with`, then you don't have to deal with it in three other locations.